### ISSUE TITLE
How to insert Array value into field?

### ISSUE THREAD

**Issue by lordent** [kayak/pypika#143]
For example
```sql
CREATE TABLE public.catalog (
   id           SERIAL PRIMARY KEY 
  ,name         VARCHAR(100) NOT NULL
  ,price        decimal(6,2) NOT NULL
  ,retail_price decimal(6,2) NOT NULL
  ,code         VARCHAR(100) NOT NULL
  ,property_values   integer[]
  ,images	VARCHAR[]
);

```
```python
    sql = str(
        pypika.PostgreSQLQuery
        .into(pypika.Table('catalog', schema='public'))
        .insert(
            (
               1,
               'test',
               '0.00',
               '0.00',
               'test',
               [1, 2, 3],
               ["some", "cool"],
            )
        )
    )
```

**Comment by lordent** [comment]
My temporary solution

```python
class Array(terms.Term):
    def __init__(self, value, alias=None):
        super().__init__(alias=alias)
        self.value = value

    def value_to_sql(self, value):
        if isinstance(value, (list, set, tuple)):
            return self.array_to_sql(self.value)
        if isinstance(value, str):
            return '\'%s\'' % value.replace('\'', '\'\'')
        return str(value)

    def array_to_sql(self, array):
        return '[%s]' % ','.join(map(self.value_to_sql, array))

    def get_sql(self, quote_char=None, **kwargs):
        return alias_sql('ARRAY' + self.array_to_sql(self.value), self.alias, quote_char)
```

**Comment by twheys** [comment]
hi @lordent 

If this isn't working it's a bug. There is an implementation already which should handle array inputs. Tuples are also supported. I'll take a look. Thanks for the examples.

**Comment by lordent** [comment]
@twheys 
```python

class Array(terms.Term):

    def __init__(self, value, alias=None):
        super().__init__(alias=alias)
        self.value = value

    def fields(self):
        return []

    def value_to_sql(self, value):
        if isinstance(value, (list, set, tuple)):
            return self.array_to_sql(self.value)
        if isinstance(value, str):
            return '\'%s\'' % value.replace('\'', '\'\'')
        return str(value)

    def array_to_sql(self, array):
        return '[%s]' % ','.join(tuple(self.value_to_sql(v) for v in array))

    def get_sql(self, quote_char=None, **kwargs):
        return alias_sql(
            'ARRAY' + self.array_to_sql(self.value)
            if self.value else 'DEFAULT',
            self.alias,
            quote_char
        )

class ArrayComparator(enums.Comparator):
    contains = '@>'
    isContainedBy = '<@'
    overlap = '&&'

# examples:
creterion = (
            terms.BasicCriterion(
                ArrayComparator.contains, 
                catalog_table.property_values, 
                Array(values)
            )
        )
        query = (
            query
            .where(creterion)
        )

    sql = str(
        pypika.PostgreSQLQuery
        .into(pypika.Table('catalog', schema='public'))
        .insert(*set(
            (
               product['id'],
               product['name'] or '',
               '0.00',
               '0.00',
               product['code'] or '',
               Array(product['property_values']),
               Array(product['images'])
            ) for product in products.values()
        ))
    )
```

**Comment by lordent** [comment]
tuples are not the same as arrays https://www.postgresql.org/docs/9.1/static/functions-array.html

**Comment by lordent** [comment]
@twheys It will be great if you add operators to work with arrays

**Comment by twheys** [comment]
So I played around with this a little bit...

```
sql = str(
        PostgreSQLQuery
        .into(Table('catalog', schema='public'))
        .insert(
            (
               1,
               'test',
               '0.00',
               '0.00',
               'test',
               [1, 2, 3],
               ["some", "cool"],
            )
        )
    )
print(sql)
```

Currently pypika converts this array into a tuple, which is what it supports already:

`INSERT INTO "public"."catalog" VALUES (1,'test','0.00','0.00','test',(1,2,3),('some','cool'))`

I'm guessing the syntax doesn't work with psql arrays, right? Unfortunately I don't have access to a psql instance to test. 

We don't need a special array class for the insert function, a python list would be good enough. The insert function wraps raw data types with a wrapper. Currently, these arrays are getting wrapped as a tuple. The tricky part is knowing when it's supposed to be an array syntax and when it should be a tuple, without forcing the users to always use a list or tuple when it's appropriate. However, maybe that's not a bad option either.

**Comment by lordent** [comment]
@twheys tuples is not a same as arrays
```
test1=# INSERT INTO "public"."catalog" VALUES (1,'test','0.00','0.00','test',(1,2,3),('some','cool'));
ERROR:  column "property_values" is of type integer[] but expression is of type record
LINE 1: ..."."catalog" VALUES (1,'test','0.00','0.00','test',(1,2,3),('...
                                                             ^
HINT:  You will need to rewrite or cast the expression.
test1=# INSERT INTO "public"."catalog" VALUES (1,'test','0.00','0.00','test',array[1,2,3],array['some','cool']);
test1=# INSERT INTO "public"."catalog" VALUES (2,'test','0.00','0.00','test',array[1,2,3],array['some','cool']);
INSERT 0 1
```

**Comment by twheys** [comment]
Right, just wanted to confirm what I was seeing. I'll change it to render lists as arrays and tuples as tuples.

**Comment by twheys** [comment]
published in version 0.15.1, could you please check your use case and let me know if you have any additional issues?
